### DISCUSSION TITLE
MAUI Android laggy performance when moving map

### DISCUSSION THREAD

**Discussion by angelinn** [Mapsui/Mapsui#2255] [Q&A] [answered]
Hello,

I've been working on a MAUI app for some time, that includes a map. After deploying the app on the Android device, when I try to move it, it feels very laggy - like it's skipping frames. Comparing it to a famous maps app you can see a big difference.

To exclude the chance that my app is making the map laggy, I built a fresh MAUI app with just the map as described in the "Getting started" guide. The result is the same - low fps when moving or zooming.

From what I remember when I had this app on Xamarin, I had such performance issues on Android, but not on UWP. Currently I can't test the MAUI app on UWP.

The device I'm testing on is Realme 9 Pro running android 12.

Has anyone encountered similar behavior on android and do you have any suggestions on how to improve the performance?

**Comment by janusw** [comment]
What kind of map do you have? Is it just a plain raster map, or are painting any additional icons/geometries on top of it?

Maybe you can check if you see the same problem with the samples provided in the Mapsui repo (so that we have a reproducible case)? https://github.com/Mapsui/Mapsui/tree/master/Samples

**Comment by angelinn** [comment]
I am using the standard sample from https://mapsui.com/documentation/getting-started-maui.html, creating the map like this:
```
public MainPage()
{
  InitializeComponent();
  
  var mapControl = new Mapsui.UI.Maui.MapControl();
  mapControl.Map?.Layers.Add(Mapsui.Tiling.OpenStreetMap.CreateTileLayer());
  Content = mapControl;
}
```

I tried to build the sample app, but at the moment it requires preview version of visual studio that I can't install.

**Reply by pauldendulk** [reply]
You should use the develop/4.1 branch to build the sample that corresponds to v4.1 (`git checkout develop/4.1`), it does not need the preview.

**Reply by pauldendulk** [reply]
btw, the default sample you definitely not be laggy. We should investigate what is going on. Perhaps it is a problem specific to your device. It would be great if you have another device to test with.

**Comment by pauldendulk** [comment]
Btw, the Blazor samples can also be opened in a mobile browser. It is a new platform and does not perform great on mobile yet, but on my phone it performance okayish. Not fast, but I would not call it laggy. I used Chrome, the browser could make a difference. How does its performance compare to your MAUI Android app?
[http:](https://mapsui.com/samples/)

**Comment by pauldendulk** [comment]
Here a video of Mapsui.Samples.Maui on an RealMe C2. Started from Visual Studio in Debug mode. Is this what you see?

https://github.com/Mapsui/Mapsui/assets/963462/4b6c474d-56ea-4982-8c41-ea3e518a3bac

**Reply by angelinn** [reply]
I have run the sample app and the performance is better - same as in your video. Making a fresh MAUI app and adding mapsui from nuget causes sluggishness. I have to look for what might be different. Will make a video to show you soon.

**Reply by pauldendulk** [reply]
Things that could be related:
- Does it make a difference if the debugger is connected? (running from Visual Studio)
- Does xaml reload make a difference?
- Is it laggy right from the start or after a while?

**Reply by angelinn** [reply]
* I deploy only in release, without debugger, I believe this will be the most optimized for such a test
* It is laggy from the beginning, yes

**Comment by angelinn** [comment]
Hello, I did a lot of tests through the last time I posted, but did not get a clear answer to myself what exactly is going on. I'll try to lay it out as good as I can. 

The sluggish behavior that I am encountering feels like frame skipping - the movement when panning/zooming is not smooth and feels rough - running on low frame rate. 

Observations:
* In all my tests on different devices ranging from a potato to a S23 ultra, the map is stuttering when panned/zoomed - this excludes a hardware issue.
* All the tests were built in Release without a debugger attached
* The problem cannot be seen when taken a video of - the camera makes everything look the same. If we compare we should test out with apks.
* The same application ran on Xamarin.Forms does not stutter when panned/zoomed - something about MAUI may be the culprit
* Creating a new MAUI project on .net8/mapsui 4.1.8 and putting the sample code from https://mapsui.com/documentation/getting-started-maui.html with nothing else causes the map to be skipping frames
* Downloading the source code from the main branch as of today and running the sample app MapsUI.Samples.Maui results in 2 different outcomes:
    * if I run the app as is, with the samples dropdown - the portion of the map that's visible has notable improvement and feels smoother to pan/zoom
    * if i run with setting the samples dropdown widthrequest to 0 to make it disappear - the map takes all of the screen and it feels like it's skipping frames again, like in the previous example
    
   My conclusion is that migrating to MAUI caused a rendering problem, due to unknown reason for me. Either that, or there's some hardware acceleration to be turned on manually. 

Expected performance of a OSM map would be https://play.google.com/store/apps/details?id=org.osmdroid&hl=en which runs smooth on every move.
   Let me know if I can do anything to provide more information or tests, as I am interested in improving the performance to be comparable to the one of osmdroid.

**Reply by charlenni** [reply]
For me it sounds like a CPU/GPU problem. See comment [here](https://github.com/Mapsui/Mapsui/blob/9d66cccfe87bf13936a6055ef6cbe0461ae0a5db/Mapsui.UI.Maui/MapControl.cs#L24). MAUI still works with CPU rendering, which isn't very fast. If you increse the number of pixels that have to be drawn, then you need the CPU to do this. When I'm remebering it correct, then Forms was capable of using the GPU.

Have you tested to reduce the size of the map control or the details you want to draw? Do you have the same problems, if you use SkiaSharp for MAUI directly and increase the number of things to draw?

**Reply by inforithmics** [reply]
https://github.com/Mapsui/Mapsui/issues/2810
 It sounds for me like this problem. The reason is that the Map of the Mapcontrol wasn’t properly initialized. When you already set the Map then this isn’t the solution.

**Reply by angelinn** [reply]
Drawing a 100x100 map seems to drastically improve performance. Might be the CPU then.

**Reply by charlenni** [reply]
Did you try to use the performance widget? There you could see some timings that could be interesting in your case.

**Reply by angelinn** [reply]
I have not, but might do it to see actual numbers.

**Comment by inforithmics** [comment]
Another thing that can be tried is setting MapControl.UseGpu = true;
This enables GPUs rendering for Android it’s disabled by default. With the Skiasharp 3 Update it will be enabled by default.
https://github.com/Mapsui/Mapsui/pull/2603

**Reply by angelinn** [reply]
That was my first thought as well, as I read in the issues linked in the code comments of [charlenni's link](https://github.com/Mapsui/Mapsui/blob/9d66cccfe87bf13936a6055ef6cbe0461ae0a5db/Mapsui.UI.Maui/MapControl.cs#L24) that the GPU acceleration of Skia has been fixed. 
Turning the `UseGpu` flag to `true` though causes an exception `Microsoft.Maui.Platform.HandlerNotFoundException: 'Handler not found for view SkiaSharp.Views.Maui.Controls.SKGLView.'` 

From what I see in the new Mapsui 5.0 source `UseSkiaSharp()` does not have a `registerRenderers` parameter. After decompiling `UseSkiaSharp()` from MapsUI 4 and 5, seems that in 5 it does not register the `SKGLView` renderer.

**Reply by angelinn** [reply]
More info: https://github.com/mono/SkiaSharp/issues/2838

**Comment by angelinn** [comment, marked as answer]
If I go back to develop/4.1 and put `Mapsui.UI.Maui.MapControl.UseGPU = true;` the performance noticably changes comparable to the one in Xamarin.Forms - a lot smoother. 

Although, If I change view to another one and then go back to the map, the map disappears and there's only a white background. This confirms GPU works on Android in MapsUI 4.1, but there's something else causing this map disappearing.

**Reply by angelinn** [reply]
Found a workaround, if you change the height of the map, it will reappear. 
Dirty hack:
```
        protected override void OnNavigatedTo(NavigatedToEventArgs args)
        {
            base.OnNavigatedTo(args);
            if (map != null)
                map.HeightRequest -= 1;
        } 
```

**Reply by inforithmics** [reply]
Sounds for me like this pull request would fix that.
https://github.com/mono/SkiaSharp/pull/3076

**Reply by janusw** [reply]
> Found a workaround

As an alternative, toggling the map visibility should also work. That is the workaround I use ...

```
        protected override void OnAppearing()
        {
#if ANDROID
            theMap.IsVisible = false;
            theMap.IsVisible = true;
#endif
            base.OnAppearing();
        }
```

See also https://github.com/mono/SkiaSharp/issues/2550. And yes, I'm also looking forward to this being fixed (via https://github.com/mono/SkiaSharp/pull/3076) 😃

**Reply by angelinn** [reply]
Works well and is more clean, thanks.

How should MapsUI proceed about this issue? Android supports GPU, but you need the workaround. I assume removing the check for Android has to happen after skia have fixed the view disappearing. 

Should we write an issue in MapsUI about this?

**Reply by inforithmics** [reply]
I have checked in a Workaround in the SkiaSharp 3 pull request

https://github.com/Mapsui/Mapsui/pull/2603/commits/836249f3320a84d743fa6e247cd9bd8e3edb9a63

**Reply by inforithmics** [reply]
And here is the a sperate Pull request for the fix

https://github.com/Mapsui/Mapsui/discussions/2255

**Comment by pauldendulk** [comment]
This should be fixed in 5.0.0-beta.6. Please test.
